### ISSUE TITLE
Add SoftPool algorithm to the SPP module

### ISSUE THREAD

**Issue by Leprechault** [ultralytics/yolov5#9622]
Please, I'd like to know the changes that I need to proceed to add the SoftPool algorithm to the SPP module. My objective is to enhance pooling operations and retain more feature information for small objects.

**Comment by glenn-jocher** [comment]
@Leprechault you can modify the SPPF() module used in YOLOv5 here:
https://github.com/ultralytics/yolov5/blob/2373d5470e386a0c63c6ab77fbee6d699665e27b/models/common.py#L214-L230

**Comment by Leprechault** [comment]
@glenn-jocher do you agree with:

```
import torch.nn.functional as F
from torch.nn.modules.utils import _triple, _pair, _single
```

```
s 214 to 230 in [2373d54](https://github.com/ultralytics/yolov5/commit/2373d5470e386a0c63c6ab77fbee6d699665e27b)

class SoftPool2d(nn.Module):
    def __init__(self, kernel_size, stride, padding=0):
        super(SoftPool2d,self).__init__()
        self.kernel_size = kernel_size
        self.stride = stride
        self.padding = padding

    def forward(self, x):
        x = self.soft_pool2d(x, kernel_size=self.kernel_size, stride=self.stride)
        return x

    def soft_pool2d(self, x, kernel_size=2, stride=None, force_inplace=False):
        kernel_size = _pair(kernel_size)
        if stride is None:
            stride = kernel_size
        else:
            stride = _pair(stride)
        _, c, h, w = x.size()
        e_x = torch.sum(torch.exp(x),dim=1,keepdim=True)
        return F.avg_pool2d(x.mul(e_x), kernel_size, stride=stride).mul_(sum(kernel_size)).div_(F.avg_pool2d(e_x, kernel_size, stride=stride).mul_(sum(kernel_size)))

def soft():
    print("You are using Soft Pooling Method")
    return SoftPool2d
```

**Comment by glenn-jocher** [comment]
@Leprechault seems interesting. Does this provide any accuracy improvements?

**Comment by Leprechault** [comment]
@glenn-jocher I'll test yet accuracy and mAP in my data set. But some papers mention this modification as a highly significant improvement.  I'll return soon.

**Comment by glenn-jocher** [comment]
@Leprechault got it. So the idea is to replace MaxPool2d() in SPFF() with SoftPool2d()?

@AyushExel @Laughing-q

**Comment by Leprechault** [comment]
> @Leprechault got it. So the idea is to replace MaxPool2d() in SPFF() with SoftPool2d()?
> 
> @AyushExel @Laughing-q 

Yeh @glenn-jocher , I'd like replace Max to Soft.

**Comment by glenn-jocher** [comment]
@Leprechault I get this error when training YOLOv5s with SoftPool2d():

<img width="770" alt="Screenshot 2022-09-28 at 19 09 35" src="https://user-images.githubusercontent.com/26833433/192845960-cef5bdc0-c59c-4384-a939-e58bcfec1f91.png">

**Comment by glenn-jocher** [comment]
This op errors:

<img width="1156" alt="Screenshot 2022-09-28 at 19 11 39" src="https://user-images.githubusercontent.com/26833433/192846219-42f7aa9c-7495-4f63-b436-28810b041f96.png">

**Comment by glenn-jocher** [comment]
It looks like kernel size is larger than grid, producing the error.

**Comment by glenn-jocher** [comment]
I've used SoftPool2d() with the same exact arguments as MaxPool2d():
<img width="651" alt="Screenshot 2022-09-28 at 19 13 39" src="https://user-images.githubusercontent.com/26833433/192846623-307327df-9726-4ebf-aba5-caf53759cac4.png">

**Comment by glenn-jocher** [comment]
I've added padding argument to both Avgpool2d calls and now this seems to train error free:

```python
class SoftPool2d(nn.Module):
    def __init__(self, kernel_size, stride, padding=0):
        super(SoftPool2d, self).__init__()
        self.kernel_size = kernel_size
        self.stride = stride
        self.padding = padding

    def forward(self, x):
        x = self.soft_pool2d(x, kernel_size=self.kernel_size, stride=self.stride)
        return x

    def soft_pool2d(self, x, kernel_size=2, stride=None):
        kernel_size = _pair(kernel_size)
        stride = kernel_size if stride is None else _pair(stride)
        e_x = torch.sum(torch.exp(x), dim=1, keepdim=True)
        return F.avg_pool2d(x.mul(e_x), kernel_size, stride=stride, padding=self.padding).mul_(sum(kernel_size)).div_(
            F.avg_pool2d(e_x, kernel_size, stride=stride, padding=self.padding).mul_(sum(kernel_size)))
```

**Comment by Leprechault** [comment]
Thanks @glenn-jocher, nice solution!!

**Comment by Leprechault** [comment]
> I've added padding argument to both Avgpool2d calls and now this seems to train error free:
> 
> ```python
> class SoftPool2d(nn.Module):
>     def __init__(self, kernel_size, stride, padding=0):
>         super(SoftPool2d, self).__init__()
>         self.kernel_size = kernel_size
>         self.stride = stride
>         self.padding = padding
> 
>     def forward(self, x):
>         x = self.soft_pool2d(x, kernel_size=self.kernel_size, stride=self.stride)
>         return x
> 
>     def soft_pool2d(self, x, kernel_size=2, stride=None):
>         kernel_size = _pair(kernel_size)
>         stride = kernel_size if stride is None else _pair(stride)
>         e_x = torch.sum(torch.exp(x), dim=1, keepdim=True)
>         return F.avg_pool2d(x.mul(e_x), kernel_size, stride=stride, padding=self.padding).mul_(sum(kernel_size)).div_(
>             F.avg_pool2d(e_x, kernel_size, stride=stride, padding=self.padding).mul_(sum(kernel_size)))
> ```

Thanks @glenn-jocher, nice solution!!

**Comment by Leprechault** [comment]
anks @glenn-jocher, nice solution!!

**Comment by ghkdtkddl** [comment]
@glenn-jocher 
did soft pooling work in detecting objects smaller than max pooling?

**Comment by glenn-jocher** [comment]
@ghkdtkddl The integration of SoftPool2d into the SPP module is still experimental. We're currently evaluating its performance on various datasets. Theoretically, SoftPool should retain more feature information, which could potentially improve detection of smaller objects. However, we'll need to complete our tests and review the results before we can make any definitive claims. Stay tuned for updates on our findings! If you have any results from your own tests, we'd love to hear about them. 🧪📊

**Comment by ghkdtkddl** [comment]
@glenn-jocher  tanks!
